### ISSUE TITLE
Unclear how to capture unhandled exceptions in callbacks

### ISSUE THREAD

**Issue by yjbanov** [flutter/flutter#10761]
Consider an `onPressed` callback that invokes an asynchronous function, such as `MethodChannel.invokeMethod` (or pretty much any plugin):

```dart
new RaisedButton(
  child: new Text('Click me!'),
  onPressed: () async {
    final channel = const MethodChannel('crashy-custom-channel');
    await channel.invokeMethod('blah');
  },
),
```

If `invokeMethod` throws an error, it will not be forwarded to `FlutterError.onError`. But also it will not be forwarded to the `Zone` where `runApp` was run, because the framework does not preserve zones, always kicking you out to the root zone.

Example: the following will not catch the error from `onPressed`, letting it fall through to the `Zone.ROOT` default handler, which simply prints it to STDOUT:

```dart
void main() {
  FlutterError.onError = (FlutterErrorDetails details) {
    print('From handler 1: $details');
  };

  runZoned<Future<Null>>(() async {
    runApp(new CrashyApp());
  }, onError: (error, stackTrace) {
    print('From handler 2: $error');
  });
}
```

We need to learn to capture _all_ Dart exceptions in order to offer a reliable crash reporting solution.

**Comment by Hixie** [comment]
Yeah, FlutterError.onError is only about reporting flutter framework errors (errors that are caught by the framework typically during a frame).

To catch the error above you can wrap it in a try-catch block? ;-)

Filed https://github.com/dart-lang/sdk/issues/29907.

**Comment by yjbanov** [comment]
As of https://github.com/flutter/flutter/commit/a4a276b71a620d6d11b95426725737437e19dbdf the `runZoned` approach works.

**Comment by Levi-Lesches** [comment]
I am having trouble with this: 

```dart
void main() async {
  FlutterError.onError = (_) => print ("Flutter error caught");
  runZoned<Future<void>>(
    () async {
      runApp(MyApp())
    },
    onError: (error, stack) => print ("Error caught by me"),
  );
}

class MyApp() extends StatelessWidget {
  @override
  Widget build(BuildContext context) => Scaffold(
    floatingActionButton: FloatingActionButton(
      child: Icon(Icons.error),
      onPressed: () async {
        await Future<void>.delayed(Duration(seconds: 1));
        String temp;
        print (temp.isEmpty);
    ),
  );
}
```

And I get: 
```
[ERROR:flutter/lib/ui/ui_dart_state.cc(148)] Unhandled Exception: NoSuchMethodError: The getter 'isEmpty' was called on null.
E/flutter (17576): Receiver: null
E/flutter (17576): Tried calling: isEmpty
```

**Comment by Levi-Lesches** [comment]
I also tried specifying a [ZoneSpecification](https://api.dartlang.org/stable/2.5.1/dart-async/ZoneSpecification-class.html), which didn't fire the `handleUncaughtErrors` or `errorCallback` functions either

**Comment by Levi-Lesches** [comment]
Ok, I haven't exactly tested this, but it seems #8742 is the issue here. 

TL;DR: Apparently, debug mode is trigger-happy when it comes to catching errors without rethrowing them.

**Comment by HumbleCarl** [comment]
@Levi-Lesches you're printing `temp.isEmpty` but you forgot that you declare it as a string and also you didn't put value on it.
